### PR TITLE
Fix salt-cloud sync_after_install ImmutableDict error

### LINKED ISSUE #61946: [BUG] Can't get custom grains sync at provisioning (via salt-cloud)
**Description**
When provisioning windows VM I'm trying to get a custom grains sync'ed somehow but cannot get it to work.

**Setup**
I've been trying to get using : 
- salt-cloud config file, hitting [this](https://github.com/saltstack/salt/issues/54452)
- orchestration state + reactor :
```
# Reactor entry :
- 'salt/cloud/*/deploy_windows':
    - '/srv/reactors/windows_first_start.sls'

# in /srv/reactors/windows_first_start.sls : 
sync_grains:
  local.saltutil.sync_grains:
    - tgt: {{ data['name'] }}

windows_orch:
  runner.state.orchestrate:
    - mods: orch.windows
    - pillar:
        vm_name: {{ data.name }}

# Orchestration states, in orch.windows : 
{% set vm_name = salt.pillar.get('vm_name') %}

wait_waht_for:
  cmd.run:
    - name: sleep 20
 
##### Sync_all to make custom modules available
#####+ https://github.com/saltstack/salt/issues/54452
#
load_my_custom_module:
  module.run:
    - name: saltutil.sync_all
    - refresh: True

orchestrate_windows_first_start:
  salt.state:
    - tgt: '{{ vm_name }}'
    - sls:
      - common.windows
```

Please be as specific as possible and give set-up details.
- [x] VM as QEMU hosted on proxmox

  
**Steps to Reproduce the behavior**
provision windows QEMU on proxmox host and try to get custom grains available after provisioning

**Expected behavior**
Getting custom grains available when running orchestration states straight after the provisioning

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

######### Master's version : 

```yaml
Salt Version:
          Salt: 3004.1
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.5
     gitpython: 2.1.11
        Jinja2: 2.10
       libgit2: 1.1.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: 2.6.1
  pycryptodome: 3.6.1
        pygit2: 1.6.1
        Python: 3.7.3 (default, Jan 22 2021, 20:04:44)
  python-gnupg: Not Installed
        PyYAML: 3.13
         PyZMQ: 17.1.2
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.1
 
System Versions:
          dist: debian 10 buster
        locale: utf-8
       machine: x86_64
       release: 5.4.162-1-pve
        system: Linux
       version: Debian GNU/Linux 10 buster
```

######### Minion's version : 
```yaml
Salt Version:
          Salt: 3004.1

Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: 4.0.7
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: 1.1.4
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.10.1
        pygit2: Not Installed
        Python: 3.8.8 (tags/v3.8.8:024d805, Feb 19 2021, 13:08:11) [MSC v.1928 32 bit (Intel)]
  python-gnupg: 0.4.7
        PyYAML: 5.4.1
         PyZMQ: 19.0.0
         smmap: 4.0.0
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.2

System Versions:
          dist:
        locale: cp1252
       machine: AMD64
       release: 2021
        system: Windows
       version: 2021 10.0.19041 SP0 Multiprocessor Free
```
</details>

**Additional context**
Logs from the master while provisioning, I can see the sync_all module is ran but nothing is found to be sync'ed : 
```                "module_|-load_my_custom_module_|-saltutil.sync_all_|-run": {
                    "__id__": "load_my_custom_module",
                    "__run_num__": 1,
                    "__sls__": "orch.windows",
                    "changes": {
                        "ret": {
                            "beacons": [],
                            "clouds": [],
                            "engines": [],
                            "executors": [],
                            "grains": [],
                            "log_handlers": [],
                            "matchers": [],
                            "modules": [],
                            "output": [],
                            "pillar": [],
                            "proxymodules": [],
                            "renderers": [],
                            "returners": [],
                            "sdb": [],
                            "serializers": [],
                            "states": [],
                            "thorium": [],
                            "utils": []
                        }
                    },
                    "comment": "Module function saltutil.sync_all executed",
                    "duration": 1732.087,
                    "name": "saltutil.sync_all",
                    "result": true,
                    "start_time": "07:45:24.661448"
                },
```

Custom grains is available to the master : 
```
root@salt-master-01:~# salt-run fileserver.file_list  
- _grains/presentusersgrain.py
[...]
```

Right after the provisioning I can see there is no present_users grains using grains.ls ...

And I can even sync after the vm is actually provisioned :  
```
root@salt-master-01:~# salt gitlab-runner-win-01  saltutil.sync_grains
gitlab-runner-win-01:
    - grains.presentusersgrain
```
And see it then : 
```
root@salt-master-01:~# salt gitlab-runner-win-01 grains.ls
gitlab-runner-win-01:
    - biosversion
    - cpu_model
    - cpuarch
    - cwd
    - disks
    - domain
    - efi
    - efi-secure-boot
    - fqdn
    - fqdn_ip4
    - fqdn_ip6
    - fqdns
    - gpus
    - groupname
    - host
    - hwaddr_interfaces
    - id
    - init
    - ip4_interfaces
    - ip6_interfaces
    - ip_interfaces
    - ipv4
    - ipv6
    - kernel
    - kernelrelease
    - kernelversion
    - locale_info
    - localhost
    - manufacturer
    - master
    - mem_total
    - motherboard
    - nodename
    - num_cpus
    - num_gpus
    - os
    - os_family
    - osfinger
    - osfullname
    - osmanufacturer
    - osrelease
    - osrelease_info
    - osservicepack
    - osversion
    - path
    - pending_reboot
    - pid
    - present_users
    - productname
    - ps
[...]
```

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
Fixes the following error and adds support for other sync functions.
```
[ERROR   ] There was a query error: 'ImmutableDict' object has no attribute 'update'
```

### What issues does this PR fix or reference?
Fixes: #61946

### Previous Behavior
Any attempts to run `sync_after_install` would fail due to attempting to update an immutable dictionary.

### New Behavior
A copy of the default master options is used instead of the source immutable dict.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/pytests/unit/cloud/test_cloud.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    master_config["sync_after_install"] = "all"
+    cloud = Cloud(master_config)
+    cloud.clouds["test.create"] = lambda x: True
+
+    with patch(
+        "salt.client.get_local_client",
+        MagicMock(return_value=MagicMock(return_value=True)),
+    ):
+        ret = cloud.create(vm_config)
+    assert ret
```
Comment (Ch3LL) [inline tests/pytests/unit/cloud/test_cloud.py:38]: can we also verify that we called the `saltutil.sync_{}` command? I'm just trying to ensure we also ensure we are hitting the part of the code that we intend and call the correct function.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/cloud/test_cloud.py (hunk 1 above)
Comment (nicholasmhughes) [reply tests/pytests/unit/cloud/test_cloud.py:38]: What's the best way to accomplish that? I tried messing around with patching the client, but the cloud code is inside a `with` statement (line 1268) so the `MagicMock` object is closed before it's passed back to the test code. This makes it impossible to check `assert_called`, `assert_called_with`, etc. on return.  @waynew , any test magic tricks?

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-m2crypto-pytest

**Review by Ch3LL — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-m2crypto-pytest

**Pull request merged**